File: lib/features/auth/data/auth_repository.dart
import 'package:flutter_riverpod/flutter_riverpod.dart';

import '../../../core/mock/mock_network.dart';

class InvalidOtpException implements Exception {
  const InvalidOtpException();
}

abstract interface class AuthRepository {
  Future<void> sendOtp(String phone);
  Future<String> verifyOtp(String phone, String code);
}

class FakeAuthRepository implements AuthRepository {
  FakeAuthRepository(this._network);

  final MockNetwork _network;

  @override
  Future<void> sendOtp(String phone) => _network.delay();

  @override
  Future<String> verifyOtp(String phone, String code) async {
    await _network.delay();
    final isSixDigits = code.length == 6 && int.tryParse(code) != null;
    if (!isSixDigits) throw const InvalidOtpException();
    return 'user-$phone';
  }
}

final authRepositoryProvider = Provider<AuthRepository>(
  (ref) => FakeAuthRepository(ref.watch(mockNetworkProvider)),
);

File: src/features/auth/data/authRepository.ts
import type { MockNetwork } from '../../../core/mock/mockNetwork';

export class InvalidOtpException extends Error {
  constructor() {
    super('InvalidOtpException');
    this.name = 'InvalidOtpException';
  }
}

export interface AuthRepository {
  sendOtp(phone: string): Promise<void>;
  verifyOtp(phone: string, code: string): Promise<string>;
}

export class FakeAuthRepository implements AuthRepository {
  constructor(private readonly network: MockNetwork) {}

  async sendOtp(_phone: string): Promise<void> {
    await this.network.delay();
  }

  async verifyOtp(phone: string, code: string): Promise<string> {
    await this.network.delay();
    if (!/^\d{6}$/.test(code)) throw new InvalidOtpException();
    return `user-${phone}`;
  }
}
